libdevcore/StringUtils.cpp

```diff
@@ -0,0 +1,77 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file StringUtils.h
+ * @author Balajiganapathi S <balajiganapathi.s@gmail.com>
+ * @date 2017
+ *
+ * String routines
+ */
+
+#include "StringUtils.h"
+#include <algorithm>
+#include <string>
+#include <vector>
+
+using namespace std;
+using namespace dev;
+
+namespace dev
+{
+
+bool stringWithinDistance(string const& _name1, string const& _name2, size_t _maxDistance)
+{
+	if (_name1 == _name2)
+		return true;
+
+	size_t n1 = _name1.size(), n2 = _name2.size();
+	vector<vector<size_t>> dp(n1 + 1, vector<size_t>(n2 + 1));
+
+	// In this dp formulation of Damerau–Levenshtein distance we are assuming that the strings are 1-based to make base case storage easier.
```

Comment:
This seems rather complicated compared to other implementations: https://en.wikibooks.org/wiki/Algorithm_Implementation/Strings/Levenshtein_distance#C.2B.2B

Especially the two-dimensional vector could be rather heap-heavy, especially given that this routine will be called on every identifier that is in scope.

Reply:
What is the license on the wikibook examples?

Reply:
@chriseth the linked example does not consider transpositions. e.g. for "test", "tset" would be considered as distance 2 (i.e. 2 substitution). I think this is a common typo so we should consider these as 1 distance.

Reply:
"Text is available under the Creative Commons Attribution-ShareAlike License.; additional terms may apply. By using this site, you agree to the Terms of Use and Privacy Policy."

but I'm not sure copyright and licensing actually applies to such small snippets.

Reply:
@chriseth @axic 

About the implementation, the implementation currently uses O(n1.n2) space. I will optimize it to O(n1) (or O(n2)) space. It will be different from the one in the wiki link as we will be accessing previous 2 entries. So will need a vector of size (3 * n). 

I can also think of another (but ugly) optimization to avoid creating a vector of 3.n1 everytime. Fix the unidentified identifier as s1, so n1 will be same over all calls to the function. So, we can create a 3 * n1 vector once and pass it to the function. 
I think we should do the 1st optimization but avoid the 2nd as it would affect generalization of the string distance function.

PS: I will do this and all the other small changes suggested here in a few hours.